### PR TITLE
Major cleanup

### PR DESCRIPTION
**Pull request by omh1280**
As @mramato talked about offline, cleanup to reduce unnecessary architecture choices.

This includes:
- Deleting `GitHubServer` and replacing with `commentOnClosedIssue` function
- Deleting `getGoogleGroupLinks` and replacing with calling `getUniqueMatches` directly
- Removing many unit tests which tested very basic JSON functionality

### CODE REVIEW

File: lib/commentOnClosedIssue.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+var commentOnClosedIssue = function(jsonResponse, headers) {
+    if (!defined(jsonResponse) || !defined(headers)) {
+        return Promise.reject('jsonResponse is undefined');
+    }
+    var commentsUrl = jsonResponse.issue.comments_url;
+    var comments = [];
+    var linkMatches = [];
+
+    return rp.get({
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Coverage indicates most of this code is never executed in the tests.

File: lib/commentOnClosedIssue.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+var defined = Cesium.defined;
+
+/** Get comments -> regex search -> post comment
+ *
+ * @param {String} jsonResponse GitHub Response
+ * @param {Object} headers Request headers
+ * @return {Promise<http.IncomingMessage>} Response
+ */
+var commentOnClosedIssue = function(jsonResponse, headers) {
+    if (!defined(jsonResponse) || !defined(headers)) {
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: This should be a developer errors using `Check`.  The rule is that anything that would normally be a dev error throws DeveloperError like normal, only runtime errors use Promise.reject.

File: lib/commentOnClosedIssue.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
+'use strict';
+var Cesium = require('cesium');
+var Promise = require('bluebird');
+var rp = require('request-promise');
+
+var getUniqueMatch = require('./getUniqueMatch');
+
+var defined = Cesium.defined;
+
+/** Get comments -> regex search -> post comment
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Write a better description about what this function does (not how it works, but what the functionality is).

File: lib/commentOnClosedIssue.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+'use strict';
+var Cesium = require('cesium');
+var Promise = require('bluebird');
+var rp = require('request-promise');
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: I'm sure their doc did it like this, but our convention is to use `requestPromise` as the variable name.

File: lib/commentOnClosedIssue.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+var defined = Cesium.defined;
+
+/** Get comments -> regex search -> post comment
+ *
+ * @param {String} jsonResponse GitHub Response
+ * @param {Object} headers Request headers
+ * @return {Promise<http.IncomingMessage>} Response
+ */
+var commentOnClosedIssue = function(jsonResponse, headers) {
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Write this as `function commentOnClosedIssue(jsonResponse, headers) {`

File: lib/commentOnClosedIssue.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+var Promise = require('bluebird');
+var rp = require('request-promise');
+
+var getUniqueMatch = require('./getUniqueMatch');
+
+var defined = Cesium.defined;
+
+/** Get comments -> regex search -> post comment
+ *
+ * @param {String} jsonResponse GitHub Response
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: This is an Object, not a string, right?  Perhaps link to the relevant GitHub documentation that defines what the object actually is.

File: lib/commentOnClosedIssue.js (hunk 5 above)
Comment (omh1280) [inline lib/commentOnClosedIssue.js]: Generally, when should I use either style?

File: lib/commentOnClosedIssue.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * @param {Object} headers Request headers
+ * @return {Promise<http.IncomingMessage>} Response
+ */
+var commentOnClosedIssue = function(jsonResponse, headers) {
+    if (!defined(jsonResponse) || !defined(headers)) {
+        return Promise.reject('jsonResponse is undefined');
+    }
+    var commentsUrl = jsonResponse.issue.comments_url;
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: I was going to suggest that we verify `issue` and `comments_url` both exist and throw an error otherwise, but it looks like `jsonResponse.issue.comments_url` is the only variable used from the `jsonReponse` object passed into this function.  Wouldn't it be better for this function to take just the `commentsUrl` string?

File: lib/commentOnClosedIssue.js (hunk 5 above)
Comment (mramato) [inline lib/commentOnClosedIssue.js]: You almost always want to use the `function xxx()` style.  The only time you may end up using the var style is when you are declaring a local function inside of another function.

File: lib/commentOnClosedIssue.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    return rp.get({
+        uri: commentsUrl,
+        headers: headers,
+        json: true
+    }).then(function(commentsJsonResponse) {
+        comments = commentOnClosedIssue._getCommentsFromResponse(commentsJsonResponse);
+        linkMatches = getUniqueMatch(comments, commentOnClosedIssue._googleLinkRegex);
+        if (linkMatches === []) {
+            return Promise.reject('No google group links found in comments!');
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Why would this reject?  It's just a no-op isn't it?

Comment (mramato) [inline lib/commentOnClosedIssue.js]: That being said, using `console.log` to spit out a status message with relevant information is probably a good idea.

File: lib/commentOnClosedIssue.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    });
+};
+
+/** Helper to abstract GitHub API
+ *
+ * @param {Object} jsonResponse GitHub response object
+ * @return {String[]} Comments
+ */
+commentOnClosedIssue._getCommentsFromResponse = function(jsonResponse) {
+    if (!defined(jsonResponse)) {
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Why would this case ever happen?

File: lib/commentOnClosedIssue.js (hunk 7 above)
Comment (omh1280) [inline lib/commentOnClosedIssue.js]: That's how I had it--I can revert it.

My thinking was that `index.js` could stay unaware about the interior details of `jsonResponse`, and just pass it along.

**Comment by mramato** [comment]
That's all I have for this go around, but I didn't look at the tests (since I suggested a bunch of changes.)  Please bump when it's ready for another look.

File: lib/commentOnClosedIssue.js (hunk 8 above)
Comment (omh1280) [inline lib/commentOnClosedIssue.js]: That's a mistake; it should just resolve

File: lib/commentOnClosedIssue.js (hunk 7 above)
Comment (mramato) [inline lib/commentOnClosedIssue.js]: That's a good strategy, but what I would do is have `commentOnClosedIssue` do nothing but extract out the `comemnts_url` and then pas it to a `commentOnClosedIssue._implemtnation` function.  This still keeps everything in the file but decouples the function, making things easier to test.  We do this in Composer all over the place, see https://github.com/AnalyticalGraphicsInc/agi-cesium-cloud/blob/master/routes/AssetServices.js#L678 for an example.

**Comment by omh1280** [comment]
Thanks! Ready for another review

File: lib/commentOnClosedIssue.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        uri: commentsUrl,
+        headers: headers,
+        json: true
+    }).then(function(commentsJsonResponse) {
+        comments = commentsJsonResponse.map(function(commentJson) {
+            return commentJson.body;
+        });
+        linkMatches = getUniqueMatch(comments, commentOnClosedIssue._googleLinkRegex);
+        if (linkMatches === []) {
+            return Promise.resolve('No google group links found in comments!');
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: This case does not appear to be covered by the test.

File: specs/lib/commentOnClosedIssueSpec.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+var fsExtra = require('fs-extra');
+var Promise = require('bluebird');
+var requestPromise = require('request-promise');
+
+var getUniqueMatch = require('../../lib/getUniqueMatch');
+
+var commentOnClosedIssue = require('../../lib/commentOnClosedIssue');
+
+describe('commentOnClosedIssue', function() {
+    it('throws if `url` or `message` is undefined', function () {
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Make this two tests, one for each argument

File: specs/lib/commentOnClosedIssueSpec.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        expect(function() {
+            commentOnClosedIssue();
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: When testing for missing parameters, make sure all other parameters are actually passed in and also specifically pass in `undefined` for the missing parameter to improve readability.

File: specs/lib/commentOnClosedIssueSpec.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        }).toThrowError();
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Normally we would use `toThrowDeveloperError` for this, but we'll have to wait until we add the Jasmine matchers stuff to be able to do it, I only mention it so you know about it.

File: specs/lib/commentOnClosedIssueSpec.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+var fsExtra = require('fs-extra');
+var Promise = require('bluebird');
+var requestPromise = require('request-promise');
+
+var getUniqueMatch = require('../../lib/getUniqueMatch');
+
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Whitespace and alphabetize these.

File: specs/lib/commentOnClosedIssueSpec.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+            commentOnClosedIssue();
+        }).toThrowError();
+
+        expect(function() {
+            commentOnClosedIssue({});
+        }).toThrowError();
+    });
+
+    it('passes correct commentsUrl to _implementation', function() {
+        spyOn(commentOnClosedIssue, '_implementation').and.callFake(function(commentsUrl) {
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Rather than using `callFake` you can just use `expect(commentOnClosedIssue._implementation).toHaveBeenCalledWith('https://api.github.com/repos/baxterthehacker/public-repo/issues/2/comments');` below after you execute the function.

Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: You should check that headers gets passed in this test as well.

File: lib/commentOnClosedIssue.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        console.log('Found these links in the comments: ', linkMatches);
+        return requestPromise.post({
+            uri: commentsUrl,
+            headers: headers,
+            body: {
+                body: 'Please make sure to update ' + linkMatches + ' on this closed issue.\n\n__I am a bot BEEEP BOOOP__'
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: We should probably make `signature` a configurable option. instead of hardcoded `\n\n__I am a bot BEEEP BOOOP__` in every message.  You can write up an issue to do this in a different PR

File: lib/commentOnClosedIssue.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    return requestPromise.get({
+        uri: commentsUrl,
+        headers: headers,
+        json: true
+    }).then(function(commentsJsonResponse) {
+        comments = commentsJsonResponse.map(function(commentJson) {
+            return commentJson.body;
+        });
+        linkMatches = getUniqueMatch(comments, commentOnClosedIssue._googleLinkRegex);
+        if (linkMatches === []) {
```
Comment (mramato) [inline lib/commentOnClosedIssue.js]: Use `linkMatches.length === 0` instead of comparing to `[]`.

File: specs/lib/commentOnClosedIssueSpec.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        });
+
+        commentOnClosedIssue(fsExtra.readJsonSync('./specs/data/issueResponse.json'), {});
+    });
+});
+
+describe('commentOnClosedIssue._implementation', function() {
+    var commentsJson = fsExtra.readJsonSync('./specs/data/issueComments.json');
+    beforeEach(function() {
+        spyOn(requestPromise, 'get').and.returnValue(Promise.resolve(commentsJson));
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Maybe there's a good reason not to do it, but instead of spying/mocking, why not just modify `issueResponse.json` to point to have it's url be `specs/data/issueComments.json`?

File: specs/lib/commentOnClosedIssueSpec.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+        spyOn(requestPromise, 'post');
+    });
+
+    it('gets the list of comments', function() {
```
Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: This (and possible other tests in this file) are asynchronous and need to take a `done` callback and cal it when the promise resolves (or done.fail if the promise fails).

Comment (mramato) [inline specs/lib/commentOnClosedIssueSpec.js]: Shouldn't this test also verify that `requestPromise.post` gets called with the expected parameters?

File: lib/getUniqueMatch.js [hunk 20]
```diff
@@ -2,19 +2,25 @@
 var Cesium = require('cesium');
 var defined = Cesium.defined;
 
+module.exports = getUniqueMatch;
+
 /** Generic regex search over array of String. (Requires global regex)
```
Comment (mramato) [inline lib/getUniqueMatch.js]: I'm not sure what `(Requires global regex)` means here,

File: specs/lib/getUniqueMatchSpec.js [hunk 21]
```diff
@@ -22,3 +25,36 @@ describe('getUniqueMatch', function() {
...
+            'Reported here: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/fewafjdsk\n\nSeen with Chrome 59.0.3071.102. iOS version: 10.3.2. Perhaps a driver issue?\n',
+            'https://groups.google.com/forum/?hl=en#!topic/cesium-dev/test1, https://groups.google.com/forum/?hl=en#!topic/cesium-dev/test2'];
+    });
+
+    it('Finds correct number of unique links', function() {
+        expect(getUniqueMatch(comments, googleLinkRegex).length).toEqual(6);
+    });
+
+    it('Returns links intact', function() {
+        expect(getUniqueMatch([comments[2]], googleLinkRegex)).toEqual([comments[2]]);
```
Comment (mramato) [inline specs/lib/getUniqueMatchSpec.js]: Writing tests that index into arrays like this make them hard to debug.  I would make a variable for each item in the `comments` array (each with a descriptive name, for example `surroundedText` and then use the variables instead of `comments[x]`.

File: gulpfile.js [hunk 22]
```diff
@@ -30,6 +30,7 @@ gulp.task('coverage', function() {
     child_process.execSync('nyc' +
         ' --all' +
         ' --report-dir coverage/' +
+        ' -x "index.js"' +
```
Comment (mramato) [inline gulpfile.js]: We shouldn't ignore `index.js` (no need to add coverage for it right now, but we should eventually).

File: lib/getUniqueMatch.js (hunk 20 above)
Comment (omh1280) [inline lib/getUniqueMatch.js]: It requires the `g` regex flag, so it looks like `/.../g`. Otherwise it will loop forever.

File: specs/lib/commentOnClosedIssueSpec.js (hunk 18 above)
Comment (omh1280) [inline specs/lib/commentOnClosedIssueSpec.js]: `requestPromise` (or `request`?) complains about formatting if the url is "specs/data/issueComments.json"

**Comment by omh1280** [comment]
Thanks a bunch @mramato. Ready for another!

File: lib/getUniqueMatch.js (hunk 20 above)
Comment (mramato) [inline lib/getUniqueMatch.js]: Please update the description to be a little more clear, also verify the input data rather than letting an infinite loop happy, you should be able to tell if it's global or not via the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/global property.

Comment (mramato) [inline lib/getUniqueMatch.js]: I'm going to merge this so I can finish reviewing #14, you can just make this update in that branch.

**Comment by mramato** [comment]
Thanks @omh1280
